compiler/noirc_frontend/src/monomorphization/mod.rs

```diff
@@ -1019,6 +1046,83 @@ impl<'interner> Monomorphizer<'interner> {
         })
     }
 
+    // Similar to `convert_type` but returns an error if any type variable can't be defaulted.
+    fn check_type(typ: &HirType, location: Location) -> Option<MonomorphizationError> {
```

Comment:
Hmm, looking at this I would rather we only had one function (convert_type) rather than having to maintain two separate functions. The cost of converting a type is usually cheap anyway since types tend not to be deeply nested.

Reply:
Sounds good, I'll try it!

Reply:
Ah, now I remember. The reason I didn't do this is because `convert_type` does this at one point:

```rust
binding.bind(default);
```

and so if we call `convert_type` with `args` and then once more with fields, it will give an error saying you can't bind twice.

Reply:
The second time it is called it should be already bound, so we shouldn't reach the same code path though, right?

Reply:
Aaaaah... you are right! I'll push another commit

Reply:
I pushed the commit but then reverted it because that produced "unexpected type 2" for some programs. I think if `args` has a numeric constant then `convert_type` fails. In `check_type` I just ignore those cases because those types are fine. But let me know if there's another approach here.

Reply:
Let me know if there's another way around this.